hail/src/main/scala/is/hail/expr/ir/NormalizeNames.scala

```diff
@@ -80,13 +80,28 @@ class NormalizeNames(normFunction: Int => String, allowFreeVariables: Boolean =
               name
         }
         Ref(newName, typ)
+      case Recur(name, args, typ) =>
+        val newName = env.eval.lookupOption(name) match {
+          case Some(n) => n
+          case None =>
+            if (!allowFreeVariables)
+              throw new RuntimeException(s"found free variable in normalize: $name")
```

Comment:
this isn't really about free variables, right? This is more "you found a recur with no buddy TailLoop"?

We should probably always error here.

Reply:
I think I disagree. Names of loops are basically variables, and recur nodes reference a loop variable. The `allowFreeVariables` flag basically means "let me run this on a subexpression sitting inside some unknown context". The set of containing `TailLoop`s is part of that context. I don't see why we should forbid running `NormalizeNames` (or any other local transformation) on the body of a `TailLoop`.

Reply:
Ah, OK, I'm convinced. I would really like to remove the allowFreeVars stuff (it was implemented as a hack to get stuff working in the state of our compiler) but there are still places where we run it on incomplete subtrees without information about context.